IBX-9594: Describe creating a custom DAM connector (#2642)
ibexa/documentation-developer: docs/content_management/images/add_image_asset_from_dam.md

### Create transformation factory

The transformation factory maps [[= product_name =]]'s image variations to corresponding variations from Wikimedia Commons.

In `src/Connector/Dam/Transformation` folder, create the `WikimediaCommonsTransformationFactory.php` file that resembles the following example,
which implements the [`TransformationFactory` interface](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Connector-Dam-Variation-TransformationFactory.html):

```php
[[= include_file('code_samples/back_office/images/src/Connector/Dam/Transformation/WikimediaCommonsTransformationFactory.php') =]]
```

Then register the transformation factory as a service:

```yaml
[[= include_file('code_samples/back_office/images/config/services.yaml', 13, 16) =]]
```

### Register variations generator

The variation generator applies map parameters coming from the transformation factory to build a fetch request to the DAM.
The solution uses the built-in `URLBasedVariationGenerator` class, which adds all the map elements as query parameters to the request.

> For example, for an asset with the ID `Ibexa_Logo.svg`, the handler generates the Asset with [`AssetUri's URL`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Connector-Dam-AssetUri.htm#method_getPatht) equal to:

Author: ```suggestion
For example, for an asset with the ID `Ibexa_Logo.svg`, the handler generates the Asset with [`AssetUri's URL`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Connector-Dam-AssetUri.html##method_getPath) equal to:
```

It is linked to the `commons` DAM source and uses the identifier `commons`. 
The DAM search tab is registered in the `connector-dam-search` [tab group](back_office_tabs.md#tab-groups) using the `ibexa.admin_ui.tab` tag.

```yaml
[[= include_file('code_samples/back_office/images/config/services.yaml', 22, 33) =]]
```

### Create Twig template

The template defines how images that come from Wikimedia Commons are displayed.

In `templates/themes/standard/`, add the `commons_asset_view.html.twig` file that resembles the following example:

```html+twig
[[= include_file('code_samples/back_office/images/templates/themes/standard/commons_asset_view.html.twig') =]]
```

Then, register the template and a fallback template in configuration files.
Replace `<scope>` with an [appropriate value](siteaccess_aware_configuration.md) that designates the SiteAccess or SiteAccess group, for example, `default` to use the template everywhere, including the back office:

```yaml
[[= include_file('code_samples/back_office/images/config/packages/views.yaml') =]]
```

### Provide back office translation

When the image asset field is displayed in the back office, a table of metadata follows.

> This example uses new fields, so you need to provide translations for their labels.

Author: ```suggestion
This example uses new fields, so you need to provide translations for their labels, for example, in `translations/ibexa_fieldtypes_preview.en.yaml`:
```

Now, when you use the Embed block in the Page Builder, you should see a DAM Image.

For more information about block customization (defined templates, variations), see [Create custom block](4_create_a_custom_block.md).

## Extend DAM support by adding custom connector

> To extend the DAM support built into [= product_name =], you must create a custom handler and transformation factory.

Author: ```suggestion
To extend the DAM support built into [[= product_name =]], you must create a custom handler and transformation factory.
```

!!! note "Wikimedia Commons licensing"

    Before you use Wikimedia Commons assets in a production environment, ensure that you comply with their [license requirements](https://commons.wikimedia.org/wiki/Commons:Reusing_content_outside_Wikimedia#How_to_comply_with_a_file's_license_requirements).

###  Create DAM handler

This class handles searching through Wikimedia Commons for images and fetching assets.

In `src\Connector\Dam\Handler` folder, create the `WikimediaCommonsHandler.php` file that resembles the following example, which uses `search()` and `fetchAsset()` functions to query the server for images and return asset objects, respectively:

```php
[[= include_file('code_samples/back_office/images/src/Connector/Dam/Handler/WikimediaCommonsHandler.php') =]]
```

Then, in `config\services.yaml`, register the handler as a service:

```yaml
[[= include_file('code_samples/back_office/images/config/services.yaml', 9, 14) =]]
```

### Create transformation factory

> The transformation factory transforms image variations coming from Wikimedia Commons to the ones used by [= product_name =].

Reviewer: ```suggestion
The transformation factory transforms image variations coming from Wikimedia Commons to the ones used by [[= product_name =]].
```

In `src\Connector\Dam\Transformation` folder, create the `WikimediaCommonsTransformationFactory.php` file that resembles the following example:

```php
[[= include_file('code_samples/back_office/images/src/Connector/Dam/Transformation/WikimediaCommonsTransformationFactory.php') =]]
```

Then register the transformation factory as a service:

```yaml
[[= include_file('code_samples/back_office/images/config/services.yaml', 15, 20) =]]
```

### Register variations generator

The variation generator applies map parameters coming from the transformation factory to build a fetch request to the DAM.
The solution uses the built-in `URLBasedVariationGenerator` class, which adds all the map elements as query parameters of the request.

For example, the handler generates the following URL with `new AssetUri()`:

`https://commons.wikimedia.org/w/index.php?title=Special:Redirect/file/Ibexa_Logo.svg`

> When the user requests a specific variation of the image, the variation generator modifies the URL and returns it in the following form:

Reviewer: ```suggestion
When the user requests a specific variation of the image, the variation generator modifies the URL and returns it in the following form (example for "large"):
```

`https://commons.wikimedia.org/w/index.php?title=Special:Redirect/file/Ibexa_Logo.svg&width=300`

For this to happen, register the variations generator as a service:

```yaml
[[= include_file('code_samples/back_office/images/config/services.yaml', 21, 25) =]]
```

### Create Twig template for Admin UI

> The template deffines how images that come from Wikimedia Commons appear in the back office.

Reviewer: ```suggestion
The template defines how images that come from Wikimedia Commons appear in the back office.
```

In `templates/bundles/WikimediaCommonsConnector/`, add the `commons_asset_view.html.twig` file that resembles the following example:

```html+twig
[[= include_file('code_samples/back_office/images/templates/themes/standard/commons_asset_view.html.twig') =]]
```

> Then, register the template in configuration files:
> 
> ```yaml
> [[= include_file('code_samples/back_office/images/config/packages/views.yaml') =]]
> ```

Reviewer: ````suggestion
Then, register the template in configuration files:

```yaml
[[= include_file('code_samples/back_office/images/config/packages/views.yaml') =]]
```

!!! caution

    As it overrides the parameter, restate the default template.
````

### Add Wikimedia Commons connection to DAM configuration

You can now configure a connection with Wikimedia Commons under the `ibexa.system.<scope>.content.dam` key:

```yaml
ibexa:
    system:
        default:
            content:
                dam: [ commons ]
```

> Once you clear the cache, you can search for images to see whether images from the newly configured DAM are displayed correctly, including their variations.

Reviewer: Worth it a screenshot with the resulting Commons tab?

For example, the handler generates the following URL with `new AssetUri()`:

`https://commons.wikimedia.org/w/index.php?title=Special:Redirect/file/Ibexa_Logo.svg`

When the user requests a specific variation of the image, for example, "large", the variation generator modifies the URL and returns it in the following form:

`https://commons.wikimedia.org/w/index.php?title=Special:Redirect/file/Ibexa_Logo.svg&width=300`

For this to happen, register the variations generator as a service:

```yaml
[[= include_file('code_samples/back_office/images/config/services.yaml', 21, 25) =]]
```

### Create Twig template for Admin UI

The template defines how images that come from Wikimedia Commons appear in the back office.

In `templates/bundles/WikimediaCommonsConnector/`, add the `commons_asset_view.html.twig` file that resembles the following example:

```html+twig
[[= include_file('code_samples/back_office/images/templates/themes/standard/commons_asset_view.html.twig') =]]
```

Then, register the template in configuration files:

```yaml
[[= include_file('code_samples/back_office/images/config/packages/views.yaml') =]]
```

!!! caution

>     As it overrides the parameter, make sure that you define the default template, too.

Reviewer 2: I don't think it is quite clear what you mean here
